rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,239 @@
+    opacity: 1,
+  },
+});
+
+export const useSampleStyles_unstable = (state: SampleState): SampleState => {
+  const styles = useStyles();
+
+  state.root.className = mergeClasses(
+    SampleClassNames.root,
+    state.visible && styles.visible,
+    state.entering && styles.entering,
+    state.exiting && styles.exiting,
+    styles.root,
+  );
+
+  return state;
+};
+```
+
+Overriding the transition/animation on the Application side:
+
+```tsx
+import * as React from 'react';
+import { Drawer } from '@fluentui/react-drawer';
+import { makeStyles } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  customDuration: {
+    transitionDuration: '500ms',
+  },
+});
+
+export const CustomDuration = () => {
+  const styles = useStyles();
+
+  return <Drawer className={styles.customDuration} />;
+};
+```
+
+#### How it works?
+
+The hook accepts a boolean value that represents the "presence" state. It internally tracks the animation/transition and return boolean values related to its motion state and a ref that needs to be assigned to the element we are tracking.
+
+```ts
+const {
+  /*
+   * Ref used to track the target element that requires animation. It should
+   * be passed to the element responsible for performing the animation.
+   */
+  ref,
+
+  /*
+   * Determines whether the component should be displayed on the screen.
+   *
+   * This property will evaluate to `true` when the provided presence value is true.
+   * Once the provided value changes to false, the hook will monitor the transition/animation
+   * completion and subsequently set this property to false.
+   */
+  shouldRender,
+
+  /*
+   * Indicates whether the component is currently rendered and visible.
+   *
+   * This flag will be set to `true` one frame after the shouldRender value becomes true.
+   * It will be set to `false` when the specified presence value changes to false.
+   */
+  visible,
+
+  /*
+   * Indicates whether the component is currently entering the screen.
+   *
+   * This will be `true` when the element is entering the screen. This happens
+   * one frame after `visible` value becomes true.
+   * It becomes `false` when the transition/animation ends.
+   */
+  entering,
+
+  /*
+   * Whether the component is leaving the screen
+   *
+   * This property is set to `true` one frame after the `visible` prop becomes true,
+   * indicating that the element is in the process of entering the screen.
+   * It becomes `false` when the transition/animation ends.
+   * It is set to false once the transition or animation for the component has completed.
+   */
+  exiting,
+
+  /*
+   * Indicates whether the component is currently undergoing animation.
+   * This property will be set to `true` when the component is either entering or exiting.
+   */
+  animating,
+} = useMotionPresence(open);
+```
+
+Events can be provided as a second argument of the hook.:
+
+```js
+  const presence = useMotionPresence(open, {
+    onEntered: () => {
```

Comment:
I don't think these states are actually necessary. The reason react-transition-group has them is because it uses those `setState(newState, callback)` calls from class components. In this case users simply need to write effects based on the state

```ts
const { entering } = useMotionPresence();

// if you don't care about browser paint
React.useEffect(() => {
  if (entering) {
    // onEntering
  }

}, [entering])

// Runs before browser paint
useIsomorphicLayoutEffect(() => {
  if (entering) {
    // onEntering
  }

}, [entering])
```

Reply:
I removed those and added a "configuration" object for the hook.

Reply:
Makes sense. I removed them in the last iteration.